Restructure twist controller

fmessmer opened the pull request:
Thanks, @ipa-fxm-mb
I really like the new structure! It will help a lot when further extending the functionality!

I created this PR for being able to leave some comments on the code....
We can discuss the comments on Monday.

Commit:
add comments

Commit:
- Restructured augmented_solver.
- Renamed class augmented_solver to AugmentedSolver.
- Created damping_methods as classes to ease creation of dampings (and new ones).

Commit:
Split up augmented_solver.cpp into different constraint solvers: JLA constraints and unconstraint.

Commit:
Made usages of ConstraintSolverFactoryBuilder:
- Creates DampingMethod
- Creates ConstraintSolver
- Executes calculation of joint velocities.

Commit:
Made restructured changes active.
Corrected some implementation.
Activated both old and new implementation for comparison and testing purposes.

Commit:
Added validation outputs.
Added comments for doxygen generation.
Did some renaming.

Commit:
- moved enfore_limits from augmented_solver to cob_twist_controller
- Added a base case WeightedLeastNorm to constraint solvers
  Instantiated it acts like an unconstraint solver.
- Renamed JointLimitAvoidanceSolver to WLN_JointLimitAvoidanceSolver
- WLN_JointLimitAvoidanceSolver inherits from WeightedLeastNormSolver and implements calculate_weighting only. -> Solving is done by the WLN Solver.

Commit:
- moved enfore_limits from augmented_solver to cob_twist_controller
- Added a base case WeightedLeastNorm to constraint solvers
  Instantiated it acts like an unconstraint solver.
- Renamed JointLimitAvoidanceSolver to WLN_JointLimitAvoidanceSolver
- WLN_JointLimitAvoidanceSolver inherits from WeightedLeastNormSolver and implements calculate_weighting

Commit:
- Added doxygen comments
- Activated graphviz for doc generation
- Added const to method signatures to avoid undesired JntArray-Data change.

Commit:
- Removed unnecessary file

Commit:
- Take care: W^(1/2) * q_dot = weighted_pinv_J * x_dot -> One must consider the weighting!!!
- Added an octave script to verify the statement above.

Commit:
- Take care: W^(1/2) * q_dot = weighted_pinv_J * x_dot -> One must consider the weighting!!!
- Added script to check pseudo-inverse calculation.

fmessmer commented on cob_twist_controller/CMakeLists.txt:
```diff
@@ -25,8 +29,16 @@ include_directories(include ${catkin_INCLUDE_DIRS} ${EIGEN_INCLUDE_DIRS} ${oroco
+
+#add_library(damping_manipulability src/damping_methods/damping_manipulability.cpp)
+# add_dependencies(damping_manipulability damping_method_base) -> nur zu andern Paketen
+#target_link_libraries(damping_manipulability damping_base)
```
will we need this commented lines or can they be removed?

ipa-fxm-mb replied:
Removed

fmessmer commented on cob_twist_controller/CMakeLists.txt:
```diff
@@ -25,8 +29,16 @@ include_directories(include ${catkin_INCLUDE_DIRS} ${EIGEN_INCLUDE_DIRS} ${oroco
 # Declare a cpp library
+add_library(damping_methods src/damping_methods/damping_base.cpp src/damping_methods/damping_manipulability.cpp src/damping_methods/damping_constant.cpp src/damping_methods/damping_builder.cpp)
+add_library(constraint_solvers ${SRC_CS_DIR}/constraint_solver_base.cpp ${SRC_CS_DIR}/unconstraint_solver.cpp ${SRC_CS_DIR}/wln_joint_limit_avoidance_solver.cpp ${SRC_CS_DIR}/weighted_least_norm_solver.cpp ${SRC_CS_FACT_DIR}/wln_joint_limit_avoidance_solver_factory.cpp ${SRC_CS_FACT_DIR}/solver_factory.cpp ${SRC_CS_FACT_DIR}/unconstraint_solver_factory.cpp ${SRC_CS_FACT_DIR}/weighted_least_norm_s ...
+target_link_libraries(constraint_solvers damping_methods)
```
should we really have a separate cpp + header for each class?
or can we group e.g. all damping_methods src's into a cpp + header and e.g. the solvers into another cpp + header?
and then have the base classes at the top and the sub-classes below...

if we group the files, we might not need the `set()` from line 4 
if we do not group, please move the `set()` into the Build section

ipa-fxm-mb replied:
Agree. Too many files. 

Suggestion: 
Providing a file for the interface definitions (.h).
Providing a file for the abstract implementation of the interface (.cpp).
One header for all implementations.
But still several cpp files for the classes to separate implementation so it will keep up organized and clean. 

I am afraid of mixing the implementations in one cpp file e.g. for the solvers.

ipa-fxm-mb replied:
Moved set() into the Build section (removed unnecessary set)

Grouped dampings, limiters, factories. 
Like stated in the last comment provided the interface definitions in a separate file. 

Still not sure about the solvers -> for the moment leave them as they are. Later (dependent on homogenous solution and/or task augmentation) a small restructure might be reasonable.

fmessmer commented on cob_twist_controller/cfg/TwistController.cfg:
```diff
@@ -11,21 +11,27 @@ damping_method_enum = gen.enum([
                        gen.const("truncation",          int_t, 2, "truncation")],
                      "enum types for the damping_method")
```
for the damping there should be the following enum types:
- None, i.e. damping_factor = 0.0
- Constant, i.e. constant damping_factor
- Manipulability, i.e. damping_factor based on manipulability

Truncation is no damping method anymore

fmessmer commented on cob_twist_controller/cfg/TwistController.cfg:
```diff
@@ -11,21 +11,27 @@ damping_method_enum = gen.enum([
-gen.add("JLA_active",   bool_t,   0, "Joint Limit Avoidance active",  True)
-gen.add("enforce_limits",   bool_t,   0, "If 'True', normalize velocities in TwistController, else let RobotHW handle the limits",  True)
-gen.add("tolerance",    double_t, 0, "Tolerance around joint position limits where velocities are normalized [degrees]",  5, 0, 15)
+gen.add("eps",               double_t, 0, "Value for singular value threshold (used for truncation)",    0.001, 0,   1)
```
eps should be (visually) separated from the parameters used for damping

ipa-fxm-mb replied:
Added an empty line before and after "eps".

ipa-fxm-mb replied:
As discussed moved it under constraint.

fmessmer commented on cob_twist_controller/cfg/TwistController.cfg:
```diff
@@ -11,21 +11,27 @@ damping_method_enum = gen.enum([
+gen.add("constraint",        int_t,    0, "The constraint to use (edited via an enum)", 2, None, None, edit_method=constraints_enum)
 
+gen.add("enforce_limits",    bool_t,   0, "If 'True', enforce joint position limits and normalize velocities in TwistController, else let RobotHW handle the limits",  True)
+gen.add("tolerance",         double_t, 0, "Tolerance around joint position limits where velocities are normalized [degrees]",  5, 0, 15)
```
should we have a separate parameter for velocity_enforcing and joint_position_enforcing?

should we have a (boolean) parameter to toggle between "keep_direction"-scaling (i.e. scale_all_with_max_factor) and "individual"-scaling (i.e. scale_each_with_respective_factor)?

ipa-fxm-mb replied:
Added keep_direction flag to cfg to choose between the scaling. 

Currently position and velocity enforcing is done (consecutively). 

With the new structure it is easy to combine and separate limiters.

ipa-fxm-mb replied:
Split enforce_limits into separate limits cfg options in rev. 39b9b490f2

fmessmer commented on cob_twist_controller/cfg/TwistController.cfg:
```diff
@@ -11,21 +11,27 @@ damping_method_enum = gen.enum([
                      "enum types for the damping_method")
 
-gen.add("damping_method", int_t, 0, "The damping method to use (edited via an enum)", 0, 0, 2, edit_method=damping_method_enum)
+constraints_enum = gen.enum([ 
```
should we call this wln_constraints_enum as we lateron might get additional constraints when using additional tasks (extra lines for Jacobian) or constraints (using Nullspace projection)?

fmessmer commented on cob_twist_controller/include/cob_twist_controller/augmented_solver.h:
```diff
@@ -19,30 +21,11 @@
 
 
-class augmented_solver
+class AugmentedSolver
```
This is a general comment:
I never really liked the wording of `augmented_solver`/`AugmentedSolver` as it only covers one aspect (i.e. augmenting the Jacobian by rows and columns) and not the whole aspect of what we are doing here.
(Its just in there from the ancient times where everything got started...)

What about using `DifferentialKinematicsSolver`?

fmessmer commented on cob_twist_controller/include/cob_twist_controller/augmented_solver.h:
```diff
@@ -56,26 +39,41 @@ class augmented_solver
     /** CartToJnt for chain using SVD including base and various DampingMethods **/
-    virtual int CartToJnt(const KDL::JntArray& q_in, const KDL::JntArray& last_q_dot, KDL::Twist& v_in, KDL::JntArray& qdot_out, std::vector<float> limits_min, std::vector<float> limits_max, KDL::Frame &base_position, KDL::Frame &chain_base);
+    virtual int CartToJnt(const KDL::JntArray& q_in, const KDL::JntArray& last_q_dot, KDL::Twist& v_in, KDL::JntArray& qdot_out, KDL::Frame &base_position, KDL::Frame &chain_base);
+    virtual int OldCartToJnt(const KDL::JntArray& q_in, const KDL::JntArray& last_q_dot, KDL::Twist& v_in, KDL::JntArray& qdot_out, KDL::Frame &base_position, KDL::Frame &chain_base);
```
This is a general comment:
In case the new structure covers the "old" functionality, I vote for dropping the support for the old structure including the functions `calculate_weighting` and `enforce_limits` which moved to other classes in the new structure!

This would further simplify and clean-up this package...

fmessmer commented on cob_twist_controller/include/cob_twist_controller/augmented_solver_data_types.h:
```diff
@@ -0,0 +1,68 @@
+ ****************************************************************/
+
+#ifndef AUGMENTED_SOLVER_DATA_TYPES_H_
+#define AUGMENTED_SOLVER_DATA_TYPES_H_
```
these classes might need a revision in case changes were made due to my comments in the cfg file to reflect them in here, too...

ipa-fxm-mb replied:
Fixed.

fmessmer commented on cob_twist_controller/include/cob_twist_controller/cob_twist_controller.h:
```diff
@@ -99,6 +99,7 @@ class CobTwistController
 	bool base_compensation_;
 	bool base_active_;
 	bool reset_markers_;
+    double tolerance_;
```
indentation

ipa-fxm-mb replied:
Fixed.

fmessmer commented on cob_twist_controller/include/cob_twist_controller/cob_twist_controller.h:
```diff
@@ -147,8 +153,21 @@ class CobTwistController
+	 * @param The inverse kinematics solution for the joint velocity vector
+	 * @return The normalized joint velocity vector
+	 */
+	KDL::JntArray normalize_velocities(const KDL::JntArray &q_dot_ik) const;
```
we might want to rename this function e.g. to `enforce_velocity_limits()`

furthermore we might want to support the two variants scale_all_with_max_factor and scale_each_with_respective_factor (see cfg)

ipa-fxm-mb replied:
See latest latest commit 5703cdc (20th April '15).
New structure for enforcing limits -> new classes which can be concatenated and made independent.

Added keep_direction parameter to cfg to choose between all and individual.

fmessmer commented on cob_twist_controller/include/cob_twist_controller/cob_twist_controller.h:
```diff
@@ -147,8 +153,21 @@ class CobTwistController
+	 * @param new_q_dot_ik The new calculated joint velocities.
+	 * @return Joint velocities adapted to limits if exceeded.
+	 */
+	KDL::JntArray enforce_limits(const KDL::JntArray& q, const KDL::JntArray& q_dot_ik) const;
```
we might want to rename this function e.g. to `enforce_position_limits()`

furthermore we might want to support the two variants scale_all_with_max_factor and scale_each_with_respective_factor (see cfg)

ipa-fxm-mb replied:
See latest latest commit 5703cdc9cf (20th April '15). 
New structure for enforcing limits -> new classes which can be concatenated and made independent. 

Added keep_direction parameter to cfg to choose between all and individual.

fmessmer commented on cob_twist_controller/include/cob_twist_controller/cob_twist_controller_debug.h:
```diff
@@ -95,14 +95,14 @@ class CobTwistController
 	
```
once the new structure is verified and tested, we also might want to drop the *_debug code

ipa-fxm-mb replied:
Removed.

fmessmer commented on cob_twist_controller/include/cob_twist_controller/constraint_solvers/constraint_solver_factory_builder.h:
```diff
@@ -0,0 +1,65 @@
+         * @param outJntVelocities The calculated joint velocities as output reference.
+         * @return The calculated new joint velocities in (m x 1)-Matrix.
+         */
+        static int8_t calculateJointVelocities(AugmentedSolverParams &asSolverParams,
```
how about keeping the KDL wording `CartToJoint` or we just call is `solve` as it is a Solver class?

minor:
in `asSolverParams` isn't the "s" already for Solver? thus `asParams``would be enough?

ipa-fxm-mb replied:
Absolutely! 
Fixed in rev. 39b9b490f2

fmessmer commented on cob_twist_controller/include/cob_twist_controller/constraint_solvers/solvers/constraint_solver_base.h:
```diff
@@ -0,0 +1,91 @@
+        inline void setDampingFactor(double damping)
+        {
+            this->dampingFactor_ = damping;
+        }
```
would there also be a function called `setWeightingMatrix` for WLN?

ipa-fxm-mb replied:
No there is no intention to provide such a function for the constraint_solver_base. 

The constraint solver base is responsible for solving the IK problem in a more general way. 

To set / add weightings it is desired to use / inherit from WeightedLeastNormSolver class. It provides a method "calculateWeighting" where a specific weighting can be implemented. Currently there is WLN_JointLimitAvoidanceSolver which implements a weighting matrix according to Chan paper ISSN 1042-296X [Page 288](Bachelor thesis Francisco)

fmessmer commented on cob_twist_controller/src/cob_twist_controller.cpp:
```diff
@@ -124,8 +124,11 @@ bool CobTwistController::initialize()
+	p_augmented_solver_ = new AugmentedSolver(chain_, 0.001, 5);
 	
+	// Before setting up dynamic_reconfigure server: init AugmentedSolverParams with default values
+	this->initAugmentedSolverParams();
```
I think, we can skip this as DynamicReconfigure uses either its default values from the cfg or those are overwritten by parameters on the ParameterServer (loaded from a yaml to the proper namespace)

fmessmer commented on cob_twist_controller/src/cob_twist_controller.cpp:
```diff
@@ -174,20 +177,23 @@ bool CobTwistController::initialize()
 	AugmentedSolverParams params;
-	params.damping_method = config.damping_method;
+    params.damping_method = static_cast<DampingMethodTypes>(config.damping_method);
+    params.constraint = static_cast<ContraintTypes>(config.constraint);
```
indentation? tabs vs. spaces

ipa-fxm-mb replied:
Fixed.

fmessmer commented on cob_twist_controller/src/cob_twist_controller.cpp:
```diff
@@ -174,20 +177,23 @@ bool CobTwistController::initialize()
+    params.limits_min = this->limits_min_; // from cob_twist_controller init
+    params.limits_max = this->limits_max_; // from cob_twist_controller init
+
+    tolerance_ = config.tolerance;
```
indentation? tabs vs. spaces

ipa-fxm-mb replied:
Fixed.

fmessmer commented on cob_twist_controller/src/cob_twist_controller.cpp:
```diff
@@ -255,6 +286,7 @@ void CobTwistController::solve_twist(KDL::Twist twist)
 	
 	int ret_ik;
 	KDL::JntArray q_dot_ik(chain_.getNrOfJoints());
+	KDL::JntArray tmp_q_dot_ik(chain_.getNrOfJoints());
```
~~do we need this? what is it used for?~~
will be removed after OldCartToJoint is removed

ipa-fxm-mb replied:
Removed!

fmessmer commented on cob_twist_controller/src/cob_twist_controller.cpp:
```diff
@@ -341,10 +380,9 @@ void CobTwistController::solve_twist(KDL::Twist twist)
 		if(enforce_limits_)
 		{
-			///Needed for limiting the base velocities
+		    q_dot_ik = enforce_limits(last_q_, q_dot_ik);
```
indentation? tabs vs. spaces

ipa-fxm-mb replied:
Fixed.

fmessmer commented on cob_twist_controller/src/damping_methods/damping_builder.cpp:
```diff
@@ -0,0 +1,53 @@
+            db = new DampingManipulability(augmentedSolverParams, jacobianData);
+            break;
+        case CONSTANT:
+        case TRUNCATION: // Always active TODO: Remove from RQT_RECONFIGURE
```
remove TRUNCATION

ipa-fxm-mb replied:
Removed!

fmessmer commented:
I get the following SegFault when commanding twists with `base_active` enabled:

It's probably a dimension mismatch when adding columns for the base DoF's...

```
[ INFO] [1429462236.582045606, 43.040000000]: StartTracking
[ INFO] [1429462236.675248083, 43.048000000]: CobFrameTracker started WITHOUT SECURITY MONITORING
[ INFO] [1429462236.775764467, 43.072000000]: ============== START AugmentedSolver::CartToJnt ==============
[ INFO] [1429462236.780021913, 43.072000000]: Old calculation of qdot_out_vec = -2.24692e-05
-2.90437e-07
-2.24691e-05
-1.1411e-08
-2.24692e-05
1.72318e-07
-2.24692e-05
-6.14989e-08
-1.05532e-10
4.44418e-08

[ INFO] [1429462236.780260927, 43.072000000]: Old calculation needed time = 0.002928

cob_twist_controller_node: /usr/include/eigen3/Eigen/src/Core/DenseCoeffsBase.h:394: Eigen::DenseCoeffsBase<Derived, 1>::Scalar& Eigen::DenseCoeffsBase<Derived, 1>::operator()(Eigen::DenseCoeffsBase<Derived, 1>::Index) [with Derived = Eigen::Matrix<double, -1, 1>; Eigen::DenseCoeffsBase<Derived, 1>::Scalar = double; Eigen::DenseCoeffsBase<Derived, 1>::Index = long int]: Assertion `index >= 0 && index < size()' failed.
[arm_left/twist_controller-32] process has died [pid 24368, exit code -6, cmd /home/fxm/projects/indigo/care-o-bot/devel/lib/cob_twist_controller/cob_twist_controller_node __name:=twist_controller __log:=/home/fxm/.ros/log/b7bb85f4-e6b3-11e4-9b42-00271039ca78/arm_left-twist_controller-32.log].
log file: /home/fxm/.ros/log/b7bb85f4-e6b3-11e4-9b42-00271039ca78/arm_left-twist_controller-32*.log
```

`base_compensation` mode works well!

Commit:
- To enforce limits for joint positions and velocities created new classes.
- Additionally added parameter for keeping direction or not when enforcing limits.
- Therefore removed normalize_velocities and enforce_limits from cob_twist_controller. Instead the new limiter_container is used.

- Added new struct to provide cob_twist_controller params.

- Removed debug code.

Commit:
- Grouped limiters in one .h and one .cpp
- Grouped damping_methods in one .h and one .cpp
- Removed separate factories. Made SolverFactory generic by introducing template parameters.
- Made usage of boost::shared_ptr instead of own pointer handling.
- Adapted CMakeLists.txt according to changes.

- Split parameter enforce_limits into enforce_pos_limits and enforce_vel_limits

fmessmer commented on cob_twist_controller/CMakeLists.txt:
```diff
@@ -16,7 +16,7 @@ generate_dynamic_reconfigure_options(
 catkin_package(
    INCLUDE_DIRS include
-   LIBRARIES cob_twist_controller cob_twist_controller_debug augmented_solver
+   LIBRARIES cob_twist_controller augmented_solver damping_methods constraint_solvers
```
`limiters` library is missing here

ipa-fxm-mb replied:
Added.

fmessmer commented on cob_twist_controller/CMakeLists.txt:
```diff
@@ -24,29 +24,29 @@ include_directories(include ${catkin_INCLUDE_DIRS} ${EIGEN_INCLUDE_DIRS} ${oroco
+
 # Declare a cpp library
+add_library(damping_methods src/damping_methods/damping_base.cpp src/damping_methods/damping.cpp)
+add_library(constraint_solvers ${SRC_CS_DIR}/constraint_solver_base.cpp ${SRC_CS_DIR}/unconstraint_solver.cpp ${SRC_CS_DIR}/wln_joint_limit_avoidance_solver.cpp ${SRC_CS_DIR}/weighted_least_norm_solver.cpp ${SRC_C_DIR}/constraint_solver_factory_builder.cpp)
```
fine with me...let's keep the solver files as they are until we get a better idea of what the solver structure will look like after null-space-projection and task-augmentation is added

fmessmer commented on cob_twist_controller/CMakeLists.txt:
```diff
@@ -24,29 +24,29 @@ include_directories(include ${catkin_INCLUDE_DIRS} ${EIGEN_INCLUDE_DIRS} ${oroco
-target_link_libraries(augmented_solver ${orocos_kdl_LIBRARIES})
+target_link_libraries(augmented_solver constraint_solvers ${orocos_kdl_LIBRARIES})
 
 add_library(${PROJECT_NAME} src/cob_twist_controller.cpp )
```
somehow it does not fit anymore calling this library `${PROJECT_NAME}` resolving to `cob_twist_controller` as other libraries are called limiters and such....can we replace `${PROJECT_NAME}` e.g. with `twist_control`in the rest of the CMakeLists.txt?

Note: In line 19 the library is explicitly called `cob_twist_controller` without using ${PROJECT_NAME}...please change it there, too

fmessmer commented:
Will we remove the "old" code (i.e. [here](https://github.com/ipa-fxm-mb/cob_control/blob/restructure_twist_controller/cob_twist_controller/src/augmented_solver.cpp#L110) or do you still feel like keeping it?
If we remove it, we should also remove everything that is only in there because it was used by the "old" code, such as damping method "truncation" in cfg and all that...

see my [comment](https://github.com/ipa-fxm/cob_control/pull/1/files#r28652689)

fmessmer commented on cob_twist_controller/include/cob_twist_controller/augmented_solver.h:
```diff
@@ -56,26 +39,51 @@ class augmented_solver
+        x_(0.0),
+        y_(0.0),
+        r_(0.0),
+        z_(0.0)
```
I can't find any usage of `r_` in the code -->remove

I think we can also get rid of x_, y_, z_ by setting 
`Eigen::Vector3d r_base_link(base_position.p.x(), base_position.p.y(), base_position.p.z());` directly

ipa-fxm-mb replied:
Corrected it like suggested. 
Then removed unused x_, y_, z_, r_ and maxiter_.

fmessmer commented on cob_twist_controller/include/cob_twist_controller/constraint_solvers/factories/solver_factory.h:
```diff
@@ -0,0 +1,103 @@
+#include <Eigen/Core>
+#include <Eigen/SVD>
+#include <kdl/jntarray.hpp>
+// #include "cob_twist_controller/constraint_solvers/solvers/constraint_solver_base.h"
```
do we need this include? or is it obsolete?

ipa-fxm-mb replied:
Removed!

fmessmer commented on cob_twist_controller/include/cob_twist_controller/constraint_solvers/solvers/constraint_solver_base.h:
```diff
@@ -0,0 +1,90 @@
+
+        const AugmentedSolverParams &asParams_; ///< References the augmented solver parameters.
+        const Matrix6Xd &jacobianData_; ///< References the current Jacobian (matrix data only).
+        const Eigen::Transpose<Matrix6Xd> &jacobianDataTransposed_; ///< References the current Jacobian transpose (matrix data only).
```
do we need a separate member for the transposed?
how is this different from jacobianData_.transpose() (or whatever the Eigen-API for transposing a matrix is)?

Is it actually used somewhere? I'm only able to find it in function headers but never in function bodies...or am I missing something?

ipa-fxm-mb replied:
Formerly used to calculate the pseudo-inverse. 
To save computation time several calls on eigen transpose() have been removed and only once called in AugmentedSolver. 

With the new calculation by SVD it will be removed.

fmessmer commented on cob_twist_controller/src/augmented_solver.cpp:
```diff
@@ -1,309 +1,389 @@
 }
 
-Eigen::VectorXd augmented_solver::calculate_weighting(const KDL::JntArray& q, const KDL::JntArray& last_q_dot, std::vector<float> limits_min, std::vector<float> limits_max)
+int AugmentedSolver::OldCartToJnt(const KDL::JntArray& q_in, const KDL::JntArray& last_q_dot, KDL::Twist& v_in, KDL::JntArray& qdot_out, KDL::Frame &base_position, KDL::Frame &chain_base)
```
Are we ready to drop this "old" code?

ipa-fxm-mb replied:
Dropped!

fmessmer commented:
Thanks for all the effort you are putting into this!
Great job! :+1:

fmessmer commented:
As a reminder to myself:
Update parameter yaml's in `cob_robots` and `schunk_robots` according to new cfg structure (if necessary)!

Commit:
- Added debug code
- Removed truncation
- Removed unused members

Commit:
- Removed unnecessary ROS_INFO_STREAMs
- Removed temporary variables for test code

ipa-fxm-mb commented:
- Tested code and wrote test report (on smb in stud-directory)
- catkin_make install worked (0 errors)
- source install/setup.bash 
  and final smoke test worked as well.

fmessmer commented on cob_twist_controller/cfg/TwistController.cfg:
```diff
@@ -7,25 +7,32 @@ gen = ParameterGenerator()
                        gen.const("manipulability",      int_t, 0, "manipulability"),
-                       gen.const("constant",            int_t, 1, "constant"),
-                       gen.const("truncation",          int_t, 2, "truncation")],
+                       gen.const("constant",            int_t, 1, "constant"),], 
```
Can we have the following damping_methods?
None (=0) -> damping_factor = 0.0 konstant
Constant (=1) -> damping_factor = damping_factor dynamic_reconfigure param
Manipulability (=2) -> damping_factor = f(manipulability)

Then we would be consistent with wln_constraints (from disabled to more advanced technique)

Then we would

ipa-fxm-mb replied:
Added a new parameter "No_damping" (cause None cannot be used twice in config). 
And made the proposed restructure.

ipa-fxm-mb replied:
Now it is a little bit confusing because "No_damping" is selected and in the line below there is a damping factor showing a value != 0.0. 

Would it be better to update the configuration after selection. What about the following workflow?
1. Selecting "No_damping" 
2. This will update the dynamic reconfigure GUI and set the damping factor to 0.0
3. As long as "No_damping" is selected no other value than 0.0 can be entered. 
4. Selecting "Constant" would free the field damping_factor and one can enter a value != 0.0. 

Is this possible?

fmessmer commented on cob_twist_controller/cfg/TwistController.cfg:
```diff
@@ -7,25 +7,32 @@ gen = ParameterGenerator()
                      "enum types for the damping_method")
 
-gen.add("damping_method", int_t, 0, "The damping method to use (edited via an enum)", 0, 0, 2, edit_method=damping_method_enum)
+constraints_enum = gen.enum([ 
```
Can we call this wln_constraint so the meaning is clear in the AugmentedSolverParams data structure?
(Having future extensions in mind...)

ipa-fxm-mb replied:
Do not agree for now. 

Still do not know how to expand new constraints. Additionally I propose not to have too many parameters to configure. Already now it is confusing. 

Therefore my suggestion is not to change it now. With new constraints I would do that work twice then.

fmessmer commented on cob_twist_controller/src/cob_twist_controller.cpp:
```diff
@@ -33,545 +33,541 @@
+    ///initialize configuration control solver
+    p_fksolver_vel_ = new KDL::ChainFkSolverVel_recursive(chain_);    //used for debugging
+    p_augmented_solver_ = new AugmentedSolver(chain_, 0.001);
+    p_old_augmented_solver_ = new AugmentedSolver(chain_, 0.001); // only for debugging
```
Is `p_old_augmented_solver` an unintended leftover? Can it be removed?

ipa-fxm-mb replied:
Removed!

fmessmer commented on cob_twist_controller/include/cob_twist_controller/cob_twist_controller_data_types.h:
```diff
@@ -0,0 +1,57 @@
+
+#include <vector>
+#include <stdint.h>
+#include <std_msgs/Float64MultiArray.h>
```
Do you need this message header in here?

ipa-fxm-mb replied:
Removed (but cannot get information about unused includes - is this possible?)!

fmessmer commented on cob_twist_controller/src/augmented_solver.cpp:
```diff
@@ -1,309 +1,139 @@
 
-augmented_solver::~augmented_solver()
-{}
+    Eigen::MatrixXd new_qdot_out_vec;
```
rename `new_qdot_out_vec` to `qdot_out_vec``as there is no **old** anymore

ipa-fxm-mb replied:
Renamed!

fmessmer commented:
Looks very nice now! Awesome!
Have a look at my last comments, then feel free to open a PullRequest to ipa320/cob_control directly!

Hopefully we will have the chance to test it with the real robot hardware during our next Hackathon next Wednesday,

Commit:
- Added new damping method None
- Added enum value to select damping None
- Removed pure pointer usage and added boost::shared_ptr usage (which provides pointer management / ensure deletion of objects)
- Removed unused includes
- Renamings

Commit:
- Instead of creating png create eps.

fmessmer commented:
closing this one as it will be merged in https://github.com/ipa320/cob_control/pull/33
